include/zephyr/rtio/rtio_spsc.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#ifndef ZEPHYR_RTIO_SPSC_H_
+#define ZEPHYR_RTIO_SPSC_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+
+/**
+ * @brief RTIO SPSC API
+ * @defgroup rtio_spsc RTIO SPSC API
+ * @ingroup rtio
+ * @{
+ */
+
+/**
+ * @file rtio_spsc.h
+ *
+ * @brief A lock-free and type safe power of 2 fixed sized SPSC Queue using a
+ * ringbuffer and atomics to ensure coherency.
+ *
+ * This SPSC implementation works on an array which wraps using a power of two
+ * size and uses a bit mask to perform a modulus. Atomics are used to allow
+ * single-producer single-consumer safe semantics without locks. Elements are
+ * expected to be of a fixed size. The API is type safe as the underlying buffer
+ * is typed and all usage is done through macros.
+ *
+ * A spsc may be declared on a stack or statically and work as intended so long
+ * as its lifetime outlives any usage. Static declarations should be the
+ * preferred method. It is meant to be a shared object between two execution
+ * contexts (ISR and a thread for example)
+ *
+ * The SPSC is safe to produce or consume in an ISR with O(1) push/pull.
+ *
+ * The SPSC is *not* safe to produce in multiple execution contexts or consume in
+ * multiple execution contexts without additional work.
+ *
+ * Safe usage would be, where A and B are unique execution contexts:
+ * 1. ISR A producing and a Thread B consuming.
+ * 2. Thread A producing and ISR B consuming.
+ * 3. Thread A producing and Thread B consuming.
```

Comment:
"ISR A producing and ISR B consuming" is not safe?

Reply:
That is safe, and should be added. What would not be sensible would be polling for any long period in an ISR and wouldn't be expected to work anyways on non-SMP systems.